geode-apis-compatible-with-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/sortedset/ZCardDUnitTest.java

```diff
@@ -0,0 +1,123 @@
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.geode.redis.internal.executor.sortedset;
+
+import static org.apache.geode.distributed.ConfigurationProperties.MAX_WAIT_TIME_RECONNECT;
+import static org.apache.geode.test.dunit.rules.RedisClusterStartupRule.DEFAULT_MAX_WAIT_TIME_RECONNECT;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import redis.clients.jedis.HostAndPort;
+import redis.clients.jedis.Jedis;
+import redis.clients.jedis.JedisCluster;
+
+import org.apache.geode.test.awaitility.GeodeAwaitility;
+import org.apache.geode.test.dunit.rules.MemberVM;
+import org.apache.geode.test.dunit.rules.RedisClusterStartupRule;
+
+public class ZCardDUnitTest {
+
+
+  @ClassRule
+  public static RedisClusterStartupRule clusterStartup = new RedisClusterStartupRule(4);
+
+  private static final int SET_SIZE = 1000;
+  private static final String LOCAL_HOST = "127.0.0.1";
+  private static final int JEDIS_TIMEOUT =
+      Math.toIntExact(GeodeAwaitility.getTimeout().toMillis());
+
+  private static MemberVM server1;
+  private static MemberVM server2;
+  private static MemberVM server3;
+
+  private static JedisCluster jedis;
+
+
+  @BeforeClass
+  public static void classSetup() {
+    Properties locatorProperties = new Properties();
+    locatorProperties.setProperty(MAX_WAIT_TIME_RECONNECT, DEFAULT_MAX_WAIT_TIME_RECONNECT);
+
+    MemberVM locator = clusterStartup.startLocatorVM(0, locatorProperties);
+    int locatorPort = locator.getPort();
+
+    server1 = clusterStartup.startRedisVM(1, locatorPort);
+    server2 = clusterStartup.startRedisVM(2, locatorPort);
+    server3 = clusterStartup.startRedisVM(3, locatorPort);
+
+    int redisServerPort = clusterStartup.getRedisPort(1);
+
+    jedis = new JedisCluster(new HostAndPort(LOCAL_HOST, redisServerPort), JEDIS_TIMEOUT);
+  }
+
+  @Before
+  public void beforeTest() {
+    try (Jedis conn = jedis.getConnectionFromSlot(0)) {
+      conn.flushAll();
+    }
+  }
+
+  @Test
+  public void shouldReturnSizeOfOneSet() {
+    String key = "key";
+    Map<String, Double> memberScoreMap = makeMemberScoreMap("member1-");
+    jedis.zadd(key, memberScoreMap);
+
+    assertThat(jedis.zcard(key)).isEqualTo(SET_SIZE);
+  }
+
+  @Test
+  public void shouldReturnSizeOfManySets_OnDifferentServers() {
```

Comment:
How are you ensuring that these sets all end up on different servers? It's possible that the keys will hash to slots hosted on the same server, unless I'm missing something.

Reply:
This could work just for the current implementation. Each key will hash to a different slot by using the hashtag, but on my end I can't guarantee those slots are on different servers. But in this case I know they land in different servers because I added temp logs to check.

Reply:
One possible solution to guarantee there are sets stored on every server might be to create a lot of them, so the chances of one server not hosting any of the buckets to which the keys map is effectively zero:
```
    for (int i = 0; i < 1000; ++i) {
      String key = "key" + i;
      Map<String, Double> memberScoreMap = makeMemberScoreMap("member" + i);
      jedis.zadd(key, memberScoreMap);
      assertThat(jedis.zcard(key)).isEqualTo(SET_SIZE);
    }
```
You could also make the set size variable for each set, to confirm that the size returned is distinct and that there's not something weird happening where some other set's size is being returned, just to make the test a little stronger.

Reply:
Great suggestion, I've redone the test to randomize the set size and the key. I've also checked the logs and see multiple servers carrying out zcard.